JASP-Engine/JASP/R/exploratoryfactoranalysis.R

```diff
@@ -27,379 +28,315 @@
+      'factorMethod',
+      'eigenValuesBox',
+      'numberOfFactors',
+      # 'incl_screePlot',
+      'plotWidthScreePlot',
+      'plotHeightScreePlot')
+    )
+  keep <- NULL
+
+  # # States:
+  # newAnalysis <- TRUE
+  # state <- .retrieveState()
+
+  # if (!is.null(state) && !is.null(state[["analysisResults"]])) {  # is there state?
   
-  # States:
-  state <- .retrieveState()
-  analysisResults <- NULL
-  newAnalysis <- TRUE
-  # if ( ! is.null(state) && state$complete) {  # is there state?
-  
-  if ( !is.null(state)) {  # is there state?
+  # # if ( !is.null(state)) {  # is there state?
     
-    diff <- .diff(options, state$options)  # compare old and new options
+  #   # nVariable <- length(options$variables)
+  #   diff <- .diff(options,state$options)
     
-    nVariable <- length(options$variables)
-    
-    if (is.list(diff) && !diff[['rotationMethod']] && !diff[['orthogonalSelector']] && !diff[['obliqueSelector']] && !diff[['variables']] && !diff[['factorMethod']] && 
-        !diff[['eigenValuesBox']] && !diff[['numberOfFactors']]) {
-      
-      # old results can be used
-      newAnalysis <- FALSE
-      analysisResults <- state$analysisResults
-      
+  #   if (is.list(diff) && !diff[['rotationMethod']] && !diff[['orthogonalSelector']] && !diff[['obliqueSelector']] && !diff[['variables']] && !diff[['factorMethod']] && 
+  #       !diff[['eigenValuesBox']] && !diff[['numberOfFactors']]) {
+
+  #     # old results can be used
+  #     newAnalysis <- FALSE
+  #   }
+  # }
+
+  if(is.null(state[["analysisResults"]])){
+
+    state <- NULL
+    analysisResults <- NULL
+
+    if (perform == "run" && !is.null(dataset) && nrow(dataset) > 1){
+
+      error0 <- .hasErrors(dataset=dataset, perform=perform, type=c("infinity", "variance"), exitAnalysisIfErrors=TRUE)
+
+      customChecks <- list(
+        function() {
+          if (length(options$variables) > 0 && options$numberOfFactors > length(options$variables)) {
+            return(paste0("Too many factors requested (", options$numberOfFactors, ") for the amount of included variables"))
+          }
+        },
+        function(){
+          if(length(options$variables) <= 1){
+            return(paste0("Not enough variables (", length(options$variables), ") for analysis."))
+          }
+        },
+        function(){
+          if(is.null(dataset)){
+            return(paste0("Dataset empty"))
+          }else if(nrow(dataset) < 3){
+            return(paste0("Not enough valid cases (", nrow(dataset),") for analysis"))
+          }
+        },
+        # check whether all row variance == 0
+        function(){
+          if(!is.null(dataset) && nrow(dataset)>1){
+            varianceZero <- 0
+            for (i in 1:nrow(dataset)){
+              if(sd(dataset[i,], na.rm = TRUE) == 0) varianceZero <- varianceZero + 1
+            }
+            if(varianceZero == nrow(dataset)){
+              return("Data not valid: variance is zero in each row")
```

Comment:
@hummyhummy shouldn't this be covered by .hasErrors? @TimKDJ is the analysis robust to no variance in nrow -1

Reply:
This is a specific check for PCA and EFA, provided by Maarten (case 4)